File: backend/app/schemas/attendance.py
from __future__ import annotations
import uuid
from datetime import date, datetime, time

from pydantic import BaseModel, Field, model_validator

from app.models.attendance import AttendanceStatus, DayOfWeek, DayType


class ScheduleUpsert(BaseModel):
    day_of_week: DayOfWeek
    is_school_day: bool = True


class ScheduleRead(BaseModel):
    id: uuid.UUID
    school_id: uuid.UUID
    day_of_week: DayOfWeek
    is_school_day: bool
    model_config = {"from_attributes": True}


class CalendarGenerateRequest(BaseModel):
    term_id: uuid.UUID
    force: bool = False  # when True, re-evaluates existing entries against current schedule


class CalendarDayRead(BaseModel):
    id: uuid.UUID
    school_id: uuid.UUID
    date: date
    day_type: DayType
    notes: str | None
    academic_term_id: uuid.UUID | None
    is_manual_override: bool
    model_config = {"from_attributes": True}


class CalendarDayOverride(BaseModel):
    day_type: DayType
    notes: str | None = Field(default=None, max_length=300)


class CalendarRangeOverride(BaseModel):
    """Mark every generated calendar day in [start_date, end_date] with the
    same day_type/notes in one call — e.g. a week-long mid-term break,
    rather than one override_calendar_day() call per day."""
    start_date: date
    end_date: date
    day_type: DayType
    notes: str | None = Field(default=None, max_length=300)

    @model_validator(mode="after")
    def _valid_range(self) -> "CalendarRangeOverride":
        if self.end_date < self.start_date:
            raise ValueError("end_date must be on or after start_date")
        if (self.end_date - self.start_date).days > 180:
            raise ValueError("Range too large — must be 180 days or fewer.")
        return self


class AttendanceMark(BaseModel):
    student_id: uuid.UUID
    status: AttendanceStatus
    notes: str | None = None


class AttendanceMarkRequest(BaseModel):
    school_calendar_id: uuid.UUID
    class_id: uuid.UUID
    records: list[AttendanceMark]
    override_reason: str | None = None
    # None = the default whole-day roll call. A real SchoolPeriod id marks
    # that one period instead — additive, never replaces the whole-day
    # record — only accepted when School.has_period_attendance is true.
    period_id: uuid.UUID | None = None


class AttendanceRecordRead(BaseModel):
    id: uuid.UUID
    student_id: uuid.UUID
    school_calendar_id: uuid.UUID
    class_id: uuid.UUID
    period_id: uuid.UUID | None
    status: AttendanceStatus
    notes: str | None
    recorded_by_id: uuid.UUID
    recorded_at: datetime
    model_config = {"from_attributes": True}


class AttendanceSummaryRead(BaseModel):
    student_id: uuid.UUID
    term_id: uuid.UUID
    total_school_days: int
    days_present: int
    days_absent: int
    days_late: int
    days_excused: int
    days_unmarked: int = 0
    attendance_rate: float  # 0.0–100.0


class TodayStatusRead(BaseModel):
    calendar_day: CalendarDayRead | None
    is_markable: bool
    record_count: int


class StudentAbsenceSummary(BaseModel):
    student_id: uuid.UUID
    days_absent: int
    days_late: int
    attendance_rate: float  # 0.0–100.0


class ClassMarkingStatusRead(BaseModel):
    class_id: uuid.UUID
    name: str
    student_count: int
    present: int
    absent: int
    marked: bool
    class_teacher_name: str | None


# ── Bell periods (SchoolPeriod) ──────────────────────────────────────────────

class PeriodCreate(BaseModel):
    name: str = Field(max_length=50)
    day_of_week: DayOfWeek
    period_number: int
    start_time: time
    end_time: time


class PeriodUpdate(BaseModel):
    name: str | None = Field(default=None, max_length=50)
    start_time: time | None = None
    end_time: time | None = None


class PeriodRead(BaseModel):
    id: uuid.UUID
    school_id: uuid.UUID
    name: str
    day_of_week: DayOfWeek
    period_number: int
    start_time: time
    end_time: time
    model_config = {"from_attributes": True}


class PeriodCopyRequest(BaseModel):
    """Clone every period on source_day onto each of target_days — the
    realistic setup path for a school whose bell times repeat Mon–Fri,
    avoiding a hand-entry of the same periods five times over."""
    source_day: DayOfWeek
    target_days: list[DayOfWeek]


class MarkablePeriod(BaseModel):
    """One row of the period picker on the Mark Attendance page — see
    services/attendance_periods.py::list_markable_periods()."""
    period_id: uuid.UUID
    name: str
    start_time: time
    end_time: time
    subject_id: uuid.UUID
    subject_name: str
    teacher_name: str | None
    can_mark: bool
    already_marked: bool

File: frontend/src/lib/api/attendance.ts
import { api } from './client';
import type { SchoolClass } from './academic';
import type { ExcuseRequest } from './portal';
export type { ExcuseRequest } from './portal';

export type DayType =
  | 'SCHOOL_DAY' | 'PUBLIC_HOLIDAY' | 'SCHOOL_HOLIDAY'
  | 'HALF_DAY'   | 'WEEKEND'        | 'EXAM_DAY';

export type AttendanceStatus = 'PRESENT' | 'ABSENT' | 'LATE' | 'EXCUSED';

export type DayOfWeek = 'MON' | 'TUE' | 'WED' | 'THU' | 'FRI' | 'SAT' | 'SUN';

export interface ScheduleDay {
  id: string;
  day_of_week: DayOfWeek;
  is_school_day: boolean;
}

export interface CalendarDay {
  id: string;
  date: string;            // 'YYYY-MM-DD'
  day_type: DayType;
  notes: string | null;
  academic_term_id: string | null;
  is_manual_override: boolean;
}

export interface AttendanceRecord {
  id: string;
  student_id: string;
  school_calendar_id: string;
  class_id: string;
  period_id: string | null;
  status: AttendanceStatus;
  notes: string | null;
  recorded_by_id: string;
  recorded_at: string;
}

// One row of the period picker on the Mark Attendance page — additive to
// the always-available "Whole day" option, never shown at all unless the
// school has opted into period-level attendance.
export interface MarkablePeriod {
  period_id: string;
  name: string;
  start_time: string;
  end_time: string;
  subject_id: string;
  subject_name: string;
  teacher_name: string | null;
  can_mark: boolean;
  already_marked: boolean;
}

export interface AttendanceSummary {
  student_id: string;
  term_id: string;
  total_school_days: number;
  days_present: number;
  days_absent: number;
  days_late: number;
  days_excused: number;
  days_unmarked: number;
  attendance_rate: number;
}

export interface TodayStatus {
  calendar_day: CalendarDay | null;
  is_markable: boolean;
  record_count: number;
}

export interface StudentAbsenceSummary {
  student_id: string;
  days_absent: number;
  days_late: number;
  attendance_rate: number;
}

export interface ClassMarkingStatus {
  class_id: string;
  name: string;
  student_count: number;
  present: number;
  absent: number;
  marked: boolean;
  class_teacher_name: string | null;
}

export const listSchedule = (): Promise<ScheduleDay[]> =>
  api.get('/attendance/schedule').then(r => r.data);

export const upsertSchedule = (data: {
  day_of_week: string;
  is_school_day: boolean;
}): Promise<ScheduleDay> =>
  api.post('/attendance/schedule', data).then(r => r.data);

export const generateCalendar = (termId: string, force = false): Promise<CalendarDay[]> =>
  api.post('/attendance/calendar/generate', { term_id: termId, force }).then(r => r.data);

export const listCalendar = (termId: string): Promise<CalendarDay[]> =>
  api.get('/attendance/calendar', { params: { term_id: termId } }).then(r => r.data);

export const overrideCalendarDay = (calId: string, data: {
  day_type: string;
  notes?: string | null;
}): Promise<CalendarDay> =>
  api.patch(`/attendance/calendar/${calId}`, data).then(r => r.data);

// Mark every already-generated calendar day in a date range at once — e.g.
// a week-long mid-term break — instead of one overrideCalendarDay() call
// per day.
export const overrideCalendarRange = (data: {
  start_date: string;
  end_date: string;
  day_type: string;
  notes?: string | null;
}): Promise<CalendarDay[]> =>
  api.patch('/attendance/calendar/range', data).then(r => r.data);

export const markAttendance = (data: {
  school_calendar_id: string;
  class_id: string;
  records: { student_id: string; status: string; notes?: string }[];
  override_reason?: string;
  period_id?: string | null;
}): Promise<AttendanceRecord[]> =>
  api.post('/attendance/mark', data).then(r => r.data);

export const listAttendanceRecords = (
  calendarId: string,
  classId: string,
  periodId?: string | null,
): Promise<AttendanceRecord[]> =>
  api.get('/attendance/records', {
    params: { calendar_id: calendarId, class_id: classId, ...(periodId ? { period_id: periodId } : {}) },
  }).then(r => r.data);

// Periods the caller can mark attendance for on this class+day — [] when
// the school hasn't enabled period-level attendance, or the day has none.
export const getMarkablePeriods = (classId: string, calendarId: string): Promise<MarkablePeriod[]> =>
  api.get('/attendance/markable-periods', { params: { class_id: classId, calendar_id: calendarId } }).then(r => r.data);

export const getTodayStatus = (classId: string): Promise<TodayStatus> =>
  api.get('/attendance/today', { params: { class_id: classId } }).then(r => r.data);

export const getClassSummaries = (classId: string, termId: string): Promise<StudentAbsenceSummary[]> =>
  api.get('/attendance/class-summaries', { params: { class_id: classId, term_id: termId } }).then(r => r.data);

export const getAttendanceSummary = (
  studentId: string,
  termId: string,
): Promise<AttendanceSummary> =>
  api.get('/attendance/summary', { params: { student_id: studentId, term_id: termId } }).then(r => r.data);

// Classes the caller can mark attendance for — scoped to their own
// ClassTeacher assignment(s) unless they hold attendance.approve (the
// backend returns the full school-wide list for those callers instead).
export const listMyAttendanceClasses = (termId: string): Promise<SchoolClass[]> =>
  api.get('/attendance/my-classes', { params: { term_id: termId } }).then(r => r.data);

// Every visible class' marking status for one calendar day — "who's marked,
// who hasn't," same scope as listMyAttendanceClasses.
export const getMarkingStatus = (calendarId: string): Promise<ClassMarkingStatus[]> =>
  api.get('/attendance/marking-status', { params: { calendar_id: calendarId } }).then(r => r.data);

// ── Chronic-absenteeism early warning ───────────────────────────────────────

export type RiskTier = 'WATCH' | 'AT_RISK' | 'SEVERE';

export interface AtRiskStudent {
  student_id: string;
  name: string;
  class_id: string | null;
  class_name: string | null;
  present: number;
  total: number;
  rate: number;
  tier: RiskTier;
}

export const getAtRiskStudents = (termId: string): Promise<AtRiskStudent[]> =>
  api.get('/attendance/at-risk', { params: { term_id: termId } }).then(r => r.data);

// ── Guardian/student absence excuse requests (staff side) ──────────────────

export const listPendingExcuseRequests = (): Promise<ExcuseRequest[]> =>
  api.get('/attendance/excuse-requests').then(r => r.data);

export const reviewExcuseRequest = (
  requestId: string,
  data: { status: 'APPROVED' | 'REJECTED'; review_notes?: string; override_reason?: string },
): Promise<ExcuseRequest> =>
  api.patch(`/attendance/excuse-requests/${requestId}/review`, data).then(r => r.data);

// ── Trends + export ──────────────────────────────────────────────────────────

export interface AttendanceTrendPoint {
  date: string;
  present: number;
  total: number;
  rate: number;
}

export const getAttendanceTrend = (termId: string, classId?: string): Promise<AttendanceTrendPoint[]> =>
  api.get('/attendance/trends', { params: { term_id: termId, ...(classId ? { class_id: classId } : {}) } }).then(r => r.data);

export const getAttendanceExportBlob = (
  termId: string, fmt: 'csv' | 'excel' | 'pdf', classId?: string,
): Promise<Blob> =>
  api.get('/attendance/export', {
    params: { term_id: termId, fmt, ...(classId ? { class_id: classId } : {}) },
    responseType: 'blob',
  }).then(r => r.data);
